graphql_compiler/cost_estimation/int_value_conversion.py

```diff
@@ -0,0 +1,101 @@
+MAX_UUID_INT = 2**128 - 1
+
+
+DATETIME_1970 = datetime.datetime(1970, 1, 1, tzinfo=pytz.utc)
+
+
+def field_supports_range_reasoning(schema_info, vertex_class, property_field):
+    """Return whether range reasoning is supported. See module docstring for definition."""
+    return any((
+        is_uuid4_type(schema_info, vertex_class, property_field),
+        is_int_field_type(schema_info, vertex_class, property_field),
+        is_datetime_field_type(schema_info, vertex_class, property_field),
+        is_date_field_type(schema_info, vertex_class, property_field),
+    ))
+
+
+def convert_int_to_field_value(schema_info, vertex_class, property_field, int_value):
+    """Return the given integer's corresponding property field value.
+
+    See module docstring for details.
+
+    Args:
+        schema_info: QueryPlanningSchemaInfo
+        vertex_class: str, name of vertex class to which the property field belongs.
+        property_field: str, name of property field that the value refers to.
+        int_value: int, integer value which will be represented as a property field value.
+
+    Returns:
+        Any, the given integer's corresponding property field value.
+
+    Raises:
+        ValueError, if the given int_value is outside the range of valid values for the given
+        property field.
+    """
+    if is_int_field_type(schema_info, vertex_class, property_field):
+        return int_value
+    if is_datetime_field_type(schema_info, vertex_class, property_field):
+        return DATETIME_1970 + datetime.timedelta(microseconds=int_value)
+    if is_date_field_type(schema_info, vertex_class, property_field):
+        return datetime.date.fromordinal(int_value)
+    elif is_uuid4_type(schema_info, vertex_class, property_field):
+        if not MIN_UUID_INT <= int_value <= MAX_UUID_INT:
+            raise ValueError(u'Integer value {} could not be converted to UUID, as it'
+                             u' is not in the range of valid UUIDs {} - {}: {} {}'
+                             .format(int_value, MIN_UUID_INT, MAX_UUID_INT, vertex_class,
+                                     property_field))
+
+        return str(UUID(int=int(int_value)))
+    elif field_supports_range_reasoning(schema_info, vertex_class, property_field):
+        raise AssertionError(u'Could not represent int {} as {} {}, but should be able to.'
+                             .format(int_value, vertex_class, property_field))
+    else:
+        raise NotImplementedError(u'Could not represent int {} as {} {}.'
+                                  .format(int_value, vertex_class, property_field))
+
+
+def convert_field_value_to_int(schema_info, vertex_class, property_field, value):
+    """Return the integer representation of a property field value."""
+    if is_int_field_type(schema_info, vertex_class, property_field):
+        return value
+    if is_datetime_field_type(schema_info, vertex_class, property_field):
+        if value.tzinfo != pytz.utc:
```

Comment:
If the datetime is timezone-naive, `tzinfo` will be None and this will get hit. We probably want to handle that case.

Reply:
Why not enforce one standard across the codebase? I can't guarantee that converting to int and back to datetime will return an equal object if we allow 2 equivalent representations of datetime values.

Reply:
I don't see a way for us to enforce this. Timezone awareness is a property of the data source being queried, and I don't see a way for us to affect that without corrupting the data we're supposed to return. Any ideas?

Reply:
The inputs and outputs to the compiler are iso formatted strings, which are always tz-aware. Unless we change that, we have to coerce naive zones to utc. But I can see how if there is a time zone in the database, preserving it in the output might be good.

This specific piece of code is dealing with datetimes from user input, and datetimes from statistics. Since we only support utc and naive, and we're not dealing with database output, it's safe to coerce everything into utc.

Reply:
What's the plan here? Are we going to assume that the caller will have normalized the field value to a UTC datetime (if it's supposed to be a datetime)?